core/kamon-core/src/main/scala/kamon/trace/SpanPropagation.scala

```diff
@@ -417,4 +417,85 @@ object W3CTraceContext {
       override def initialValue(): Array[Byte] = Array.ofDim[Byte](256)
     }
   }
+
+  /**
+    * DataDog HTTP propagation. Based on `dd-trace-java` implementation and observations from services instrumented
+    * with the DataDog Java Agent. List of all available HTTP headers used in trace propagation.
+    *
+    * https://github.com/DataDog/dd-trace-java/blob/76e41d3d51314ed2814d0c8deac78c17bb87638e/dd-trace-core/src/main/java/datadog/trace/core/propagation/DatadogHttpCodec.java#L24..L27
+    *
+    * Only a subset of possible DataDog headers are handled.
+    */
+  object DataDog {
+    def apply(): DataDog = new DataDog
+
+    object Headers {
+      val ParentSpanId = "x-datadog-parent-id"
+      val TraceId = "x-datadog-trace-id"
+      val SamplingPriority = "x-datadog-sampling-priority"
+    }
+
+    object SamplingPriority {
+      val Sample = "1"
+      val DoNotSample = "0"
+    }
+
+    final implicit class KamonIdOps(private val id: kamon.trace.Identifier) extends AnyVal {
+      def toLongId: Long = BigInt(id.string, 16).toLong
+      def toUnsignedLongString: String = BigInt(id.string, 16).toString
+    }
+
+    /**
+      * DataDog tracing ids will be 64 bit unsigned integers.
+      * https://docs.datadoghq.com/tracing/guide/send_traces_to_agent_by_api/
+      */
+    def decodeUnsignedLongToHex(id: String): String =
+      urlDecode(id).toLong.toHexString
+  }
+
+  class DataDog extends Propagation.EntryReader[HeaderReader] with Propagation.EntryWriter[HeaderWriter] {
+    import DataDog._
+
+    override def read(reader: HeaderReader, context: Context): Context = {
+      val identifierScheme = Kamon.identifierScheme
+      val traceId = reader.read(Headers.TraceId)
+        .map(id => identifierScheme.traceIdFactory.from(decodeUnsignedLongToHex(id)))
+        .getOrElse(Identifier.Empty)
+
+      // To preserve the parent/child relationship of spans in DataDog the `x-parent-parent-id` is used as the span id
+      // of the new span. See `DatadogHttpCodec` implementation in `dd-trace-java` for details.
+      val spanId = reader.read(Headers.ParentSpanId)
+        .map(id => identifierScheme.spanIdFactory.from(decodeUnsignedLongToHex(id)))
+        .getOrElse(Identifier.Empty)
```

Comment:
I'm not super sure if this is the place to apply this logic. Have you looked at the `kamon.trace.join-remote-parents-with-same-span-id` setting? It sounds like that setting should roughly have the same effect, although parentId will be set in that case.

Reply:
Actually, I just noticed that the client side is only including the span id as the parent id. Weird, but if that's how DD does it :shrug:

Reply:
@seglo so, Datadog ends up merging the data sent by the client-side service and the server-side service into the same Span? Asking out of curiosity

Reply:
I agree, it's confusing.

> Datadog ends up merging the data sent by the client-side service and the server-side service into the same Span?

Intuitively this is what I thought would happen as well, but the DataDog trace view shows them as separate spans.

My first thought was to assign the id passed as `x-datadog-parent-id` to the Parent ID of the Kamon trace and generate a new ID for the span itself, but this broke the parent/child association in the DataDog trace view.

Here's what we see, the top span is from our frontend and the 2nd is an akka-http service using this implementation.

![datadog-trace](https://user-images.githubusercontent.com/1148412/160648728-887b651f-dceb-4c8f-98ea-2e2e5af1d242.png)

Reply:
Thanks for sharing!